### PR TITLE
Weights and biases integration tutorial

### PR DESCRIPTION
**Pull request by kentdanas**
Start of a W&B integration tutorial with Airflow based on this repo: https://github.com/astronomer/airflow-wandb-demo/tree/simple

### CODE REVIEW

**Comment by mpgreg** [comment]
I created a separate repo for the longer version.  Link at top should go to https://github.com/astronomer/airflow-wandb-cosmos-demo

**Comment by mpgreg** [comment]
The simple branch has been moved to main.  https://github.com/astronomer/airflow-wandb-demo

File: code-samples/dags/airflow-weights-and-biases/weights_and_biases.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        model_params = model.get_params()
+
+        y_pred = model.predict(X_test)
+        y_probas = model.predict_proba(X_test)
+        importances = model.feature_importances_
+        indices = np.argsort(importances)[::-1]
+
+        wandb.login()
+        run = wandb.init(
+            project=wandb_project,
```
Comment (morganmcg1) [inline code-samples/dags/airflow-weights-and-biases/weights_and_biases.py:118]: I'd suggest adding an `entity` option here too to be able to log to a W&B Team :)

File: code-samples/dags/airflow-weights-and-biases/weights_and_biases.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        return {'wandb_project': wandb_project, 
+                'run_id': run.id, 
+                'artifact_name': model_artifact_name}
+
+    @aql.dataframe()
+    def predict(model_info: dict, customer_df: pd.DataFrame) -> pd.DataFrame:
+
+        wandb.login()
+        run = wandb.init(
+            project=model_info['wandb_project'],
```
Comment (morganmcg1) [inline code-samples/dags/airflow-weights-and-biases/weights_and_biases.py:157]: I'd suggest adding an `entity` option here too to be able to log to a W&B Team :)

**Comment by kentdanas** [comment]
@jwitz new ML-related integration tutorial for Learn. You'll see a few places where I need to fill in some info, but generally this is ready to get the editing ball rolling.

@mpgreg if you don't mind reviewing for ML-related language and any needed DAG changes based on WandB's feedback like we talked about, that would be awesome!

File: learn/airflow-weights-and-biases.md [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+    scikit-learn==1.2.2
+    ```
+
+    This installs the packages needed to transform the data and complete feature engineering, model training and predictions.
+
+## Step 2: Prepare the data
+
+This tutorial will create a model that classifies churn risk based on customer data.
+
+1. Create a folder called `data` in your `include` directory.
```
Comment (jwitz) [inline learn/airflow-weights-and-biases.md:66]:
```suggestion
1. Create a subfolder called `data` in your Astro project `include` folder.
```

File: learn/airflow-weights-and-biases.md [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+2. Download the CSV files from [this GitHub directory](https://github.com/astronomer/airflow-wandb-demo/tree/simple/include/data).
```
Comment (jwitz) [inline learn/airflow-weights-and-biases.md:67]:
```suggestion
2. Download the demo CSV files from [this GitHub directory](https://github.com/astronomer/airflow-wandb-demo/tree/simple/include/data).
```

File: learn/airflow-weights-and-biases.md [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+3. Save the downloaded CSV files in the `include/data` folder. You should have 8 total files.
```
Comment (jwitz) [inline learn/airflow-weights-and-biases.md:68]:
```suggestion
3. Save the downloaded CSV files in the `include/data` folder. You should have 8 files in total.
```

Comment (jwitz) [inline learn/airflow-weights-and-biases.md:68]:
```suggestion
3. Save the downloaded CSV files in the `include/data` folder. You should have 8 files in total.
```

Comment (jwitz) [inline learn/airflow-weights-and-biases.md:68]:
```suggestion
3. Save the downloaded CSV files in the `include/data` folder. You should have 8 files in total.
```

File: learn/airflow-weights-and-biases.md [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+
+## Step 3: Create your SQL transformation scripts
+
+Before feature engineering and training, the data needs to be transformed. This tutorial uses the Astro Python SDK `transform_file` function to complete several transformations using SQL.
+
+1. Create a file in your `include` directory called `customer_churn_month.sql` and copy the following code into the file.
```
Comment (jwitz) [inline learn/airflow-weights-and-biases.md:74]:
```suggestion
1. Create a file in your `include` folder called `customer_churn_month.sql` and copy the following code into the file.
```

File: learn/airflow-weights-and-biases.md [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+            last_active_month,
+            false as is_first_month,
+            false as is_last_month
+        from customer_revenue_by_month
+        where is_last_month
+    )
+    select * from joined1;
+    ```
+
+2. Create another file in your `include` directory called `customers.sql` and copy the following code into the file.
```
Comment (jwitz) [inline learn/airflow-weights-and-biases.md:159]:
```suggestion
2. Create another file in your `include` folder called `customers.sql` and copy the following code into the file.
```

File: learn/airflow-weights-and-biases.md [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+    from final
+    ```
+
+## Step 4: Create a W&B API Key
+
+In your W&B account, create an API key that you will use to connect Airflow to W&B. You can create a key by going to the [authorize](https://wandb.ai/authorize) page, or under your user settings.
+
+## Step 5: Set up your connections and environment variables
+
+This tutorial uses environment variables to create Airflow connections to Snowflake and W&B and configure the Astro Python SDK.
```
Comment (jwitz) [inline learn/airflow-weights-and-biases.md:216]:
```suggestion
You'll use environment variables to create Airflow connections to Snowflake and W&B, as well as to configure the Astro Python SDK.
```

Comment (jwitz) [inline learn/airflow-weights-and-biases.md:216]:
```suggestion
You'll use environment variables to create Airflow connections to Snowflake and W&B, as well as to configure the Astro Python SDK.
```

File: learn/airflow-weights-and-biases.md [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+
+1. Open the `.env` file in your Astro project and paste the following code.
+
+    ```text
+    AIRFLOW_CONN_SNOWFLAKE_DEFAULT='snowflake://<USER_NAME>:<PASSWORD>@/<ROLE_NAME>?account=<ACCOUNT_NAME>&region=<REGION_NAME>&database=<DATABASE_NAME>&warehouse=<WAREHOUSE_NAME>&role=<ROLE>'
+
+    ####WANDB
+    WANDB_API_KEY='xxxxxxxxxxxxxxxxxxxxxxx'
```
Comment (jwitz) [inline learn/airflow-weights-and-biases.md:224]:
```suggestion
    WANDB_API_KEY='<your-wandb-api-key>'
```

File: learn/airflow-weights-and-biases.md [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+
+    ###SDK
+    AIRFLOW__ASTRO_SDK__XCOM_STORAGE_URL='s3://local-xcom'
+    AIRFLOW__ASTRO_SDK__XCOM_STORAGE_CONN_ID='minio_local'
+    ```
+
+2. Update the `AIRFLOW_CONN_SNOWFLAKE_DEFAULT` variable with your Snowflake credentials.
+
+3. Update the `WANDB_API_KEY` variable with the API key you created in [Step 4](#step-4-create-a-wb-api-key).
```
Comment (jwitz) [inline learn/airflow-weights-and-biases.md:233]:
```suggestion
3. Replace `<your-wandb-api-key>` with the API key you created in [Step 4](#step-4-create-a-wb-api-key).
```

File: learn/airflow-weights-and-biases.md [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+        left join customer_payments on customers.customer_id = customer_payments.customer_id
+    )
+    select
+    *
+    from final
+    ```
+
+## Step 4: Create a W&B API Key
+
+In your W&B account, create an API key that you will use to connect Airflow to W&B. You can create a key by going to the [authorize](https://wandb.ai/authorize) page, or under your user settings.
```
Comment (jwitz) [inline learn/airflow-weights-and-biases.md:212]:
```suggestion
In your W&B account, create an API key that you will use to connect Airflow to W&B. You can create a key by going to the [Authorize](https://wandb.ai/authorize) page or your user settings.
```

File: learn/airflow-weights-and-biases.md [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+
+    ####WANDB
+    WANDB_API_KEY='xxxxxxxxxxxxxxxxxxxxxxx'
+
+    ###SDK
+    AIRFLOW__ASTRO_SDK__XCOM_STORAGE_URL='s3://local-xcom'
+    AIRFLOW__ASTRO_SDK__XCOM_STORAGE_CONN_ID='minio_local'
+    ```
+
+2. Update the `AIRFLOW_CONN_SNOWFLAKE_DEFAULT` variable with your Snowflake credentials.
```
Comment (jwitz) [inline learn/airflow-weights-and-biases.md:231]:
```suggestion
2. Update the placeholder values in `AIRFLOW_CONN_SNOWFLAKE_DEFAULT` with your Snowflake credentials.
```

**Review by jwitz — APPROVED** [review]
Only minor suggestions from me, LGTM after we have the missing content!

File: learn/airflow-weights-and-biases.md [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+import CodeBlock from '@theme/CodeBlock';
+import weights_and_biases from '!!raw-loader!../code-samples/dags/airflow-weights-and-biases/weights_and_biases.py';
+
+[Weights and Biases](https://wandb.ai/site) (W&B) is a machine learning platform for model management that includes features like experiment tracking, dataset versioning, and model performance evaluation and visualization. Using W&B with Airflow gives you a powerful ML orchestration stack with first-class features for building, training, and managing your models.
+
+In this tutorial, you'll learn how to create an Airflow DAG that completes feature engineering, model training, and predictions with the Astro Python SDK and scikit-learn, and registers the model with W&B for evaluation and visualization.
+
+:::info
+
+This tutorial was developed in partnership with Weights and Biases. For a more complex example integration that includes dbt for data transformations, see ADD LINK TO OTHER POSTS/TUTORIALS HERE.
```
Comment (kentdanas) [inline learn/airflow-weights-and-biases.md:17]:
```suggestion
This tutorial was developed in partnership with Weights and Biases. For resources on implementing other use cases with W&B, see [Tutorials](https://wandb.ai/site/tutorials).
```

File: learn/airflow-weights-and-biases.md [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+- Weights and Biases. See [What is Weights and Biases?](https://docs.wandb.ai/?_gl=1*i7pmr7*_ga*MTI3ODk4OTUzNy4xNjc5Njc2MzE5*_ga_JH1SJHJQXJ*MTY4MTI0ODQ2OS42LjEuMTY4MTI0ODUxMS4xOC4wLjA.).
+
+## Prerequisites
+
+- The [Astro CLI](https://docs.astronomer.io/astro/cli/overview).
+- A [Weights and Biases](https://wandb.ai/site) account. Personal accounts are available for free.
+
+## Quickstart:  
+  
+If you have a github account you can get started quickly by cloning the demo repository.
```
Comment (kentdanas) [inline learn/airflow-weights-and-biases.md:40]:
```suggestion
If you have a Github account, you can get started quickly by cloning the demo repository.
```

**Pull request merged**